### ISSUE TITLE
Remove synchronisation

### ISSUE THREAD

**Issue by lgrahl** [saltyrtc/saltyrtc-client-java#114]
This library is not thread-safe and thus does not need synchronisation.

**Comment by dbrgn** [comment]
I'd explore 3, potentially using a single thread executor service: https://github.com/TakahikoKawasaki/nv-websocket-client#connect-to-server-asynchronously

**Comment by ovalseven8** [comment]
To be honest, I am not entirely sure how you'd structure it with a single thread executor service? Perhaps you can elaborate.

Some other idea (please feedback):

- Encapsulate the whole WebSocket session and handling into separate object. So, when you would like to change the websocket client library in the future, it's less of a hassle. Currently, the SaltyRTC session and the websocket session is tightly coupled and kind of inflexible for changes.
- Encapsulate the whole SaltyRTC session state in one object instance. At the moment, the state is shared between multiple threads and in combination with the callbacks I find it hard to follow the code. Could be my inexperience, though. :smile: 
- What I have overall in mind is more like actor-based concurrency and message passing instead of the manual locking. 

Let's say we have two main actors (`SaltyRTC` and `WebSocket`). Both have their own mailbox for messages (just use a `BlockingQueue`). Now, we can ensure that the currently multi-threaded websocket library will nonetheless process the frames sequentially. In the callback methods we pass messages to the `SaltyRTC` and `WebSocket` actors. A little bit like [Akka](https://akka.io/), just way more lightweight and way less powerful.

It would mean to create immutable messages for message passing. Each actor would run in its own thread and executing tasks that are in the mailbox sequentially.

**Comment by ovalseven8** [comment]
I'm not entirely sure what you mean, @dbrgn. Can you explain how you would structure that?

**Comment by ovalseven8** [comment]
For the time being #125 basically introduces @lgrahl's 1./2.
